modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/core/useIE11Compatibility.js

```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {useLayoutEffect} from 'react';
+
+export default function useIE11Compatibility() {
+	useLayoutEffect(() => {
+		if (Liferay.Browser?.isIe()) {
+			const oldRemoveChild = window.HTMLElement.prototype.removeChild;
+
+			window.HTMLElement.prototype.removeChild = function () {
+				try {
+					return oldRemoveChild.apply(this, arguments);
+				}
+				catch (error) {
+					if (
+						process.env.NODE_ENV === 'development' &&
+						!!error &&
+						!!error.message &&
+						error.message === 'NotFoundError'
+					) {
+						console.warn('IE NotFoundError handled');
+					}
+					else {
+						throw error;
+					}
+				}
+			};
+
+			return () => {
+				window.HTMLElement.prototype.removeChild = oldRemoveChild;
+			};
```

Comment:
Another thing about re-usability of hooks... This hook doesn't look "safe" to be used by multiple components at once (because each usage site is going to simultaneous try to do the monkey patch in a non-idempotent way). It is only safe if you use it in at most one place at a time. This is another hint that this probably shouldn't be a hook (and much less one that gets exported from a module).

Reply:
Related with the other comment I tend to like isolating functionality to a hook to give more meaning to the useEffect, but you are right that isn't the best use case for hooks. Do you think this would fit more as a component? Something like <IE11Compatibility /> (with a better name ofc)

Reply:
I don't think the problem is the name (which is easily changed), it's that it is:

- A hook that is never re-used.
- But which is exported from a module as though it were intended for re-use.
- And is not actually safe to re-use, because it clobbers global state.

A less strong objection to it is:

- It does _nothing_ except call another hook. It doesn't take or massage params, or compose things together. Its sole justification as a layer of indirection would be for re-use (doesn't apply here) or to give it a name (ie. to make it clear what it is for).

But if all you want to do is give it a name, a comment would do just as well without the downsides listed above; eg.

```js
// IE issue: https://issues.liferay.com/browse/LPS-XXXXXX

useLayoutEffect(() => {
    // ...
});
```

(Also, why does this have to be `useLayoutEffect`? If it's because there's some kind of race going on and that you have to patch ASAP, that's another bit of a warning sign...)

I think you could deal with some of the concerns here with something like this:

```js
// App.js
let removeChild;

/**
 * Some comment here with a link to an LPS.
 */
function safeRemoveChild() {
    try {
        return removeChild.apply(this, arguments);
    }
    catch (error) {
        // etc...
    }
}

function App() {
    if (Liferay.Browser?.isIE() {
        removeChild = window.HTMLElement.prototype.removedChild;

        window.HTMLElement.prototype.removeChild = safeRemoveChild;
    }

    useEffect(() => {
        return () => {
            if (removeChild) {
                window.HTMLElement.prototype.removeChild = removeChild;
                removeChild = undefined;
            }
        };
    }, []);
}
```

ie. no need to use a hook to set up the monkey patch (and therefore no need to use `useLayoutEffect`) because it doesn't actually depend on the component having been rendered to the DOM yet. Of course, if there is no race, then you can harmlessly move the set-up into the `useEffect` and things will look more symmetrical, which is nice.

Reply:
Ok, I think I am going for something like this :) and go on with the monkey patching given our time constraint and that the problem is only reproducible in ie11...

I'm close this pr, apply the suggestions and send this to review to one of my teammates

Thanks a lot for your feedback I appreciate it a lot! 👏